packages/peregrine/src/List/items.js

```diff
@@ -1,104 +1,112 @@
-                    render={renderItem}
-                    hasFocus={hasFocus && cursor === index}
-                    isSelected={selection.has(key)}
-                    onBlur={this.handleBlur}
-                    onClick={this.getClickHandler(key)}
-                    onFocus={this.getFocusHandler(index)}
-                />
-            );
-        });
-
-        return <Fragment>{children}</Fragment>;
-    }
-
-    syncSelection() {
-        const { selection } = this.state;
-        const { onSelectionChange } = this.props;
-
-        if (onSelectionChange) {
-            onSelectionChange(selection);
+};
+
+const reducer = (state, action) => {
+    switch (action.type) {
+        case 'REMOVE_FOCUS':
+            return {
+                ...state,
+                hasFocus: false
+            };
+        case 'SET_FOCUS':
+            return {
+                ...state,
+                hasFocus: true,
+                cursor: action.key
+            };
+        case 'UPDATE_SELECTION': {
+            const { key, selectionModel } = action;
+            return {
+                ...state,
+                selection: updateSelection(key, state.selection, selectionModel)
+            };
         }
+        default:
+            return state;
     }
-
-    handleBlur = () => {
-        this.setState(removeFocus);
-    };
-
-    getClickHandler = memoize(key => () => {
-        this.setState(updateSelection(key), this.syncSelection);
-    });
-
-    getFocusHandler = memoize(index => () => {
-        this.setState(updateCursor(index));
+};
+
+const initialState = {
+    cursor: null,
+    hasFocus: false,
+    selection: new Set()
+};
+
+const usc = fn => useCallback(fn, []);
+
+function Items(props) {
+    const {
+        getItemKey,
+        items,
+        renderItem,
+        selectionModel,
+        onSelectionChange
+    } = props;
+    const [{ cursor, hasFocus, selection }, dispatch] = useReducer(
+        reducer,
+        initialState
+    );
+    useEffect(() => {
+        onSelectionChange && onSelectionChange(selection);
+    }, [Array.from(selection).toString()]); // when ever the selection changes, make the call
+    const removeFocus = usc(() => dispatch({ type: 'REMOVE_FOCUS' }));
+    const updateSelection = usc(key =>
+        dispatch({
+            type: 'UPDATE_SELECTION',
+            key,
+            selectionModel
+        })
+    );
+    const setFocus = usc(key => dispatch({ type: 'SET_FOCUS', key }));
+    const children = items.map((item, index) => {
+        const key = getItemKey(item, index);
+        return (
+            <ListItem
+                key={key}
+                item={item}
+                itemIndex={index}
+                render={renderItem}
+                hasFocus={hasFocus && cursor === key}
+                isSelected={selection.has(key)}
+                onBlur={removeFocus}
+                onClick={updateSelection(key)}
+                onFocus={setFocus(key)}
```

Comment:
Without memoization, these callbacks will be recreated on each render. After creating the custom hook described earlier, we should work out what new props to pass to `ListItem` and have that component create the handlers itself.

Reply:
Will something like this solve our problem of recreating handlers upon re-render?

`Commit: 55df3e063fc8f78cacf3e94e727c48ffb9360575`

```
const key = getItemKey(item, index);
const updateSelection = useMemo(() => _updateSelection(key), [ key ]);
const setFocus = useMemo(() => _setFocus(key), [key]);
```